docs(operations): add WRB cutover preparation guide (#2588) (#3008)
hiero-ledger/hiero-block-node: docs/block-node/operations/preparing-your-block-node-for-wrb-cutover.md

|----------------|----------------------------------------------------------------------------------------------------------------------|-------------------------|----------------------------------------|
| `live`         | `blockNode.persistence.live.mountPath` (default `/opt/hiero/block-node/data/live`)                                   | 20% of provisioned size | Recent block stream and live state     |
| `archive`      | `blockNode.persistence.archive.mountPath` (default `/opt/hiero/block-node/data/historic`)                            | 20% of provisioned size | Compressed historic block archive      |
| `verification` | `blockNode.persistence.verification.mountPath` (default `/opt/hiero/block-node/verification`)                        | 500 MB                  | Block hash state and verification data |
| `logging`      | `blockNode.persistence.logging.mountPath` (default `/opt/hiero/block-node/logs`)                                     | 1 GB                    | Application logs                       |

> | `plugins`      | `blockNode.persistence.plugins` (not mounted by default; present only when additional plugins are explicitly loaded) | —                       | Plugin JARs                            |

Reviewer: where does the 20% of provisioned size comes from?

|              Artifact              |                                                  Location on host                                                   |                                              Why it is irreplaceable                                               |
|------------------------------------|---------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------|
| `block-node-values.yaml`           | `/etc/solo-provisioner/block-node-values.yaml`                                                                      | Cohort-specific Helm overlay generated once at handoff                                                             |
| `rsa-bootstrap-roster.json`        | Default: `/opt/hiero/block-node/node/rsa-bootstrap-roster.json` (configurable via `app.state.rsaBootstrapFilePath`) | RSA public-key roster for WRB proof verification; only needed when not using Mirror Node auto-fetch for the roster |

> | TLS bundle (cert, key, full chain) | Operator-managed; delivered at handoff                                                                              | Required to re-establish the CN-to-BN gRPC-over-TLS connection after reinstall                                     |

Reviewer: also, am not aware we are saving TLS certs on the BN atm

> | `admin_key`                        | Operator-managed key custody                                                                                        | Required to update the on-chain BN registration; losing it makes the registration unmodifiable                     |

Reviewer: what is this? I've never head of an `admin_key` 😅

Reviewer 2: This is likely bleed-over from the HIP-1137 `RegisteredNode`.

> **Caution:** Treat the `admin_key` as a production signing key from day one. Losing it means the on-chain BN registration cannot be updated.

---

## Release-specific checks

Complete the subsection below that matches the upgrade you are preparing for. Future releases will appear here as additional subsections. If no matching subsection exists for your target release, only the common checks above are required.

---

### WRB streaming cutover prep

**Applies to:** the Consensus Node release that activates Wrapped Record Block (WRB) streaming — currently scheduled for CN release 0.77.0. The exact release may change if release testing surfaces a blocker; your Hashgraph PoC will confirm the target release before the maintenance window.

For the full network cutover timeline — phases, CN-side WRB catch-up, TSS ceremony, and Jumpstart Data — see [Cutover Process and Timeline](../Cutover-Process.md).

> **What is changing:** from the cutover release onwards, Consensus Nodes cease producing record streams and begin producing Block Streams with full TSS and WRAPS signatures. Block Nodes that have been streaming "preview" blocks (incomplete hashes, pre-cutover testing streams) must be reset before the cutover so they can receive authoritative block history from the Record Block History (RBH) Block Node via backfill.

Reviewer: This is also true when coming from preview blockstream to WRB, so this upgrade should be done `--with reset`

Author: @AlfredoG87, The reset step is already documented as a required separate step before the upgrade. Could you confirm whether Solo Provisioner supports a combined `upgrade --with reset` flag? If so, what is the exact invocation? I'll add it as the recommended single-command path once confirmed.

<!-- src: solo-weaver@222c3f47:internal/blocknode/storage.go#L388-L407 (ResetStorage function) -->
<!-- src: solo-weaver@222c3f47:cmd/cli/commands/block/node/reset.go (Use: "reset"; 5-step Long description) -->

```bash
sudo solo-provisioner block node reset --profile=mainnet
```

- **Expected output:**

  ```text
  Ensuring weaver service account (weaver:2500)
  Preflight Checks
  Scaling down Block Node
  Clearing Block Node storage
  Scaling up Block Node
  Waiting for Block Node to be ready

  Completed successfully
  ```

Wait until the pod reaches `1/1 Running`, then confirm the store is empty:

```bash
grpcurl -plaintext -emit-defaults \
  -import-path ~/bn-proto \
  -proto block-node/api/node_service.proto \
  -d '{}' \
  "$BLOCK_NODE_HOST:40840" \
  org.hiero.block.api.BlockNodeService/serverStatus
```

- **Expected immediately after reset:**

  ```json
  {
    "firstAvailableBlock": "18446744073709551615",
    "lastAvailableBlock": "18446744073709551615",
    "onlyLatestState": false
  }
  ```

  Both values at `18446744073709551615` confirm the store is empty and the Block Node is ready to backfill.

> If the reset fails, check `sudo head /opt/solo/weaver/logs/solo-provisioner.log` and contact your Hashgraph PoC if the pod remains stuck.

Reviewer: would be good to also include the BN logs and the K8s pod describe.

#### Provision the RSA bootstrap roster

The `roster-bootstrap-rsa` plugin must have the RSA public-key roster at startup to verify `SignedRecordFileProof` items in incoming WRBs.

<!-- src: hiero-block-node@fa8afa83:docs/design/wrb-streaming/bootstrap-roster-plugin.md (Purpose section) -->
<!-- src: hiero-block-node@fa8afa83:block-node/roster-bootstrap-rsa/src/main/java/org/hiero/block/node/roster/bootstrap/rsa/RsaRosterBootstrapConfig.java#L24-L27 (@ConfigData roster.bootstrap.rsa, mirrorNodeBaseUrl default "") -->
<!-- src: hiero-block-node@fa8afa83:block-node/app-config/src/main/java/org/hiero/block/node/app/config/state/ApplicationStateConfig.java#L31 (rsaBootstrapFilePath default) -->

**Mirror Node auto-fetch (mainnet cohort default)**

> When `roster.bootstrap.rsa.mirrorNodeBaseUrl` is set in your `block-node-values.yaml`, the plugin fetches the roster from the Mirror Node REST API at startup and caches it locally. After `block node reset`, the cached copy is cleared along with the rest of the storage directories and is automatically re-fetched on the next pod start — no manual intervention is needed. Before the cutover window, confirm outbound connectivity to the Mirror Node is available from the BN host.

Reviewer: should we include the mirror node url for both `testnet` and `mainnet` ?

> Note: If the Mirror Node is unreachable, the plugin polls indefinitely — every 5 seconds until the first roster is fetched, then every 60 seconds for periodic refresh. Both intervals are configurable. WRB proof verification is non-functional until the roster is available.

**Peer Block Node query (optional)**

When `roster.bootstrap.rsa.blockNodeSourcesPath` is set, the plugin also queries configured peer Block Nodes via gRPC to retrieve the address book. This runs concurrently with the Mirror Node query when no bootstrap file is present; whichever responds first provides the initial roster. The peer BN query follows the same two-phase polling schedule (5-second initial interval, 60-second subsequent interval, both configurable). Configure the path to a JSON file listing your peer Block Nodes:

```json
{
  "nodes": [
    { "address": "10.0.0.1", "port": 8080, "priority": 1, "name": "peer-bn-1" },
    { "address": "10.0.0.2", "port": 8080, "priority": 2, "name": "peer-bn-2" }
  ]
}
```

> 

Reviewer: This is also required for `backfill` imo all BNs should have at least 1 BN source configured, recommended to have more than 1.

I am assuming they are referencing the same file that is created by populating the backfill yaml values.

helm -n block-node get values <release-name> | grep BACKFILL_GREEDY
```

- **Expected:** `BACKFILL_GREEDY: "true"`

**Edge case — bootstrapping from genesis when no public BN holds the history yet:**

During the initial mainnet WRB cutover, the wrapped record block history may not yet be available from public Block Nodes. In this case, a special-purpose Block Node holding the offline-wrapped WRBs serves as a temporary backfill source. Operators who need access to this node will receive the endpoint and connection details through a separate operator communication channel — it is not published in this document.

If you are directed by your Hashgraph PoC to configure a specific backfill source:

1. Update your Helm values overlay to set `BACKFILL_BLOCK_NODE_SOURCES_PATH` to the path of a `block-node-sources.json` file, then confirm it was applied:

   ```bash
   helm -n block-node get values <release-name> | grep -i backfill
   ```
2. The `block-node-sources.json` file format is a JSON array of Block Node endpoints. Use the hostname and port provided by your Hashgraph PoC:

   ```json
   [
     {
       "address": "<host-provided-by-hashgraph-poc>",
       "port": 40840
     }
   ]

>    ```

Reviewer: We should not recommend to modify the json path directly, it should be done via helm chart values overrides.

   > Note: Even when backfilling from a special-purpose BN, all blocks are cryptographically verified by the BN's verification plugin before they are stored. Block legitimacy is confirmed regardless of the backfill source.

3. Monitor backfill progress. Full backfill of current network history takes one to several weeks; the BN does not need to complete backfill before the cutover release, but it must be active and making progress:

   ```bash
   kubectl -n block-node logs $BN_POD -c block-node-server --since=5m | grep -i "backfill"

>    ```

Reviewer: Best way to monitor is by querying serverStatus or serverStatusDetail and verify that the `last_available_block` is increasing, or that the missing blocks are being filled.

   ```

   Then call `serverStatus`:

   ```bash
   grpcurl -plaintext -emit-defaults \
     -import-path ~/bn-proto \
     -proto block-node/api/node_service.proto \
     -d '{}' \
     "$BLOCK_NODE_HOST:40840" \
     org.hiero.block.api.BlockNodeService/serverStatus
   ```

   Record `firstAvailableBlock` and `lastAvailableBlock`. Both should be sensible block numbers — not `18446744073709551615` (the sentinel for "no blocks yet") — if the Block Node has been ingesting. For instructions on downloading the protobuf bundle into `~/bn-proto`, see [Connecting a Mirror Node to a Block Node](./connecting-a-mirror-node-to-a-block-node.md).

3. Confirm Alloy telemetry is shipping (if configured):

   ```bash
   kubectl -n grafana-alloy get pods
   ```

   All Alloy pods should be `Running`. A non-running Alloy pod is not a blocker for the upgrade but should be investigated afterwards.

### Validate hardware readiness

Run the Solo Provisioner hardware preflight check before the upgrade. This validates CPU core count, RAM, disk, OS version, and network-connectivity requirements for the target profile.

```bash
sudo solo-provisioner block node check --profile=mainnet
```

> - **Expected:** each step shows a green checkmark and "success".

Reviewer 2: It might help to include the exact character (I'm guessing this is the correct character):

```suggestion
- **Expected:** each step shows a green checkmark (✅) and "success".
```

|     Volume     |                                                  Default mount path                                                  | Minimum free space |                Purpose                 |
|----------------|----------------------------------------------------------------------------------------------------------------------|--------------------|----------------------------------------|
| `live`         | `blockNode.persistence.live.mountPath` (default `/opt/hiero/block-node/data/live`)                                   | 1 GB               | Recent block stream and live state     |
| `archive`      | `blockNode.persistence.archive.mountPath` (default `/opt/hiero/block-node/data/historic`)                            | 1 GB               | Compressed historic block archive      |
| `verification` | `blockNode.persistence.verification.mountPath` (default `/opt/hiero/block-node/verification`)                        | 500 MB             | Block hash state and verification data |
| `logging`      | `blockNode.persistence.logging.mountPath` (default `/opt/hiero/block-node/logs`)                                     | 1 GB               | Application logs                       |

> | `plugins`      | `blockNode.persistence.plugins` (not mounted by default; present only when additional plugins are explicitly loaded) | —                  | Plugin JARs                            |

Reviewer 2: These are local testing configuration values.
I'm not sure the goal for this document, but if it is intended for _production_ use (particularly `mainnet`), then the values for `live`, `archive`, and `logging` are much larger (6TB, 90TB, and 10GB respectively).

I think plugins is _always_ mounted (it should be...); it just doesn't necessarily have any content initially.

> | `live`         | `blockNode.persistence.live.mountPath` (default `/opt/hiero/block-node/data/live`)            | 6 TB               | Recent block stream and live state                                                      |
> | `archive`      | `blockNode.persistence.archive.mountPath` (default `/opt/hiero/block-node/data/historic`)     | 90 TB              | Compressed historic block archive                                                       |
> | `verification` | `blockNode.persistence.verification.mountPath` (default `/opt/hiero/block-node/verification`) | 500 MB             | Block hash state and verification data                                                  |
> | `logging`      | `blockNode.persistence.logging.mountPath` (default `/opt/hiero/block-node/logs`)              | 10 GB              | Application logs                                                                        |
> | `plugins`      | `blockNode.persistence.plugins`                                                               | —                  | Plugin JARs; always mounted but may contain no JARs until plugins are explicitly loaded |

Reviewer 2: Some changes are in process, but due very soon.  Also, a few of the volumes are a bit small.
Note, `blockNode.persistence.verification.mountPath` is replaced with `blockNode.persistence.applicationState.mountPath` in the next BN version.

```suggestion
| `live`         | `blockNode.persistence.live.mountPath` (default `/opt/hiero/block-node/data/live`)            | 6 TB               | Recent block stream and live state                                                      |
| `archive`      | `blockNode.persistence.archive.mountPath` (default `/opt/hiero/block-node/data/historic`)     | 90 TB              | Compressed historic block archive                                                       |
| `verification` | `blockNode.persistence.verification.mountPath` (default `/opt/hiero/block-node/verification`) | 50 GB             | Block hash state and verification data                                                  |
| `logging`      | `blockNode.persistence.logging.mountPath` (default `/opt/hiero/block-node/logs`)              | 100 GB              | Application logs                                                                        |
| `plugins`      | `blockNode.persistence.plugins`                                                               | —                  | Plugin JARs; always mounted but may contain no JARs until plugins are explicitly loaded |
```

|-----------------------------|---------------------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------|
| `block-node-values.yaml`    | `/etc/solo-provisioner/block-node-values.yaml`                                                                      | Cohort-specific Helm overlay generated once at handoff                                                             |
| `rsa-bootstrap-roster.json` | Default: `/opt/hiero/block-node/node/rsa-bootstrap-roster.json` (configurable via `app.state.rsaBootstrapFilePath`) | RSA public-key roster for WRB proof verification; only needed when not using Mirror Node auto-fetch for the roster |

---

## Release-specific checks

Complete the subsection below that matches the upgrade you are preparing for. Future releases will appear here as additional subsections. If no matching subsection exists for your target release, only the common checks above are required.

---

### WRB streaming cutover prep

> **Applies to:** the Consensus Node release that activates Wrapped Record Block (WRB) streaming — currently scheduled for CN release 0.77.0. The exact release may change if release testing surfaces a blocker; your Hashgraph PoC will confirm the target release before the maintenance window.

Reviewer 2: This is scheduled for `0.75`.  `0.77` is the cutover to _Block Streams and TSS_.

```suggestion
**Applies to:** the Consensus Node release that activates Wrapped Record Block (WRB) streaming — currently scheduled for CN release 0.75.0. The exact release may change if release testing surfaces a blocker; your Hashgraph PoC will confirm the target release before the maintenance window.
```

For the full network cutover timeline — phases, CN-side WRB catch-up, TSS ceremony, and Jumpstart Data — see [Cutover Process and Timeline](../Cutover-Process.md).

> **What is changing:** from the cutover release onwards, Consensus Nodes cease producing record streams and begin producing Block Streams with full TSS and WRAPS signatures. Any preview blocks stored by the Block Node before the cutover are invalid and must be discarded before the BN can receive and store authoritative WRB history. Do not skip the reset step even if the BN appears to be functioning normally.

Reviewer 2: This refers to TSS/Block Stream cutover, not the WRB cutover.

```suggestion
**What is changing:** from the WRB cutover release onwards, Consensus Nodes begin producing Wrapped Recordfile Blocks with aggregated RSA signature Block Proofs. Any preview blocks stored by the Block Node before the cutover are invalid and must be discarded before the BN can receive and store authoritative WRB history. Do not skip the reset step even if the BN appears to be functioning normally.
The production of Record files uploaded to S3 storage will continue until the cutover to TSS and Block Streams in a later release.
```

  -d '{}' \
  "$BLOCK_NODE_HOST:40840" \
  org.hiero.block.api.BlockNodeService/serverStatus
```

- **Expected immediately after reset:**

  ```json
  {
    "firstAvailableBlock": "18446744073709551615",
    "lastAvailableBlock": "18446744073709551615",
    "onlyLatestState": false
  }
  ```

  Both values at `18446744073709551615` confirm the store is empty and the Block Node is ready to backfill.

If the reset fails, run the following to diagnose the issue and share the output with your Hashgraph PoC:

```bash
sudo head /opt/solo/weaver/logs/solo-provisioner.log
kubectl -n block-node describe pod $BN_POD
kubectl -n block-node logs $BN_POD -c block-node-server --tail=200
```

#### Provision the RSA bootstrap roster

The `roster-bootstrap-rsa` plugin must be included in your Block Node's plugin list and have the RSA public-key roster available at startup to verify `SignedRecordFileProof` items in incoming WRBs.

**Enable the plugin**

Add `roster-bootstrap-rsa` to your plugin list in `block-node-values.yaml`. Append it to your existing `plugins.names` value — the example below shows the full plugin list used for LFH nodes on previewnet, provided for reference:

```yaml
plugins:

>   names: facility-messaging,block-access-service,health,server-status,stream-publisher,stream-subscriber,verification,blocks-file-historic,blocks-file-recent,backfill,roster-bootstrap-rsa

Reviewer 2: We should include roster-bootstrap-tss here as well, to avoid needing to restart the block node unnecessarily.

```suggestion
  names: facility-messaging,block-access-service,health,server-status,stream-publisher,stream-subscriber,verification,blocks-file-historic,blocks-file-recent,backfill,roster-bootstrap-rsa,roster-bootstrap-tss
```

    { "address": "peer2.example.com", "port": 40840, "priority": 2, "name": "peer-bn-2" }
  ]
}
```

Configure this file via your Helm chart's ConfigMap or values override mechanism. Use the endpoints provided by your Hashgraph PoC.

**Manual file (alternative)**

If `ROSTER_BOOTSTRAP_RSA_MIRROR_NODE_BASE_URL` is not configured, the plugin reads the roster from `app.state.rsaBootstrapFilePath` (default: `/opt/hiero/block-node/node/rsa-bootstrap-roster.json`). The file is delivered as part of the cohort package by your Hashgraph PoC. After `block node reset`, confirm the file is still present and non-empty:

```bash
kubectl -n block-node exec $BN_POD -c block-node-server -- \
  ls -lh /opt/hiero/block-node/node/rsa-bootstrap-roster.json
```

- **If the file is missing after reset:** re-deliver it from your off-host backup using the mechanism in your cohort's `block-node-values.yaml`.

In either case, confirm the roster loaded cleanly after the pod starts:

```bash
kubectl -n block-node logs $BN_POD -c block-node-server | grep -i "roster\|bootstrap\|rsa"
```

- **Expected:** no `WARNING` or `ERROR` lines referencing the roster.

> 

Reviewer 2: A much more reliable test is to query the Status API and read the roster data in the response.
